viz/LaserLine.cpp

```diff
@@ -2,44 +2,123 @@
 #include "LaserLine.hpp"
 
 using namespace vizkit3d;
-
-struct LaserLine::Data {
-    // Copy of the value given to updateDataIntern.
-    //
-    // Making a copy is required because of how OSG works
-    base::samples::LaserScan data;
-};
-
+using namespace vizkit3d_laser_line;
 
 LaserLine::LaserLine()
-    : p(new Data)
 {
+    color = osg::Vec4(0.0, 1.0, 0.0, 0.0);
+    vertices = new osg::Vec3Array();
+    lineHolder = new LineHolder();
+    showAllLines = false;
 }
 
 LaserLine::~LaserLine()
 {
-    delete p;
 }
 
 osg::ref_ptr<osg::Node> LaserLine::createMainNode()
 {
-    // Geode is a common node used for vizkit3d plugins. It allows to display
-    // "arbitrary" geometries
-    return new osg::Geode();
+    return lineHolder->getRootNode();
 }
 
-void LaserLine::updateMainNode ( osg::Node* node )
+void LaserLine::updateMainNode (osg::Node* node)
 {
-    osg::Geode* geode = static_cast<osg::Geode*>(node);
-    // Update the main node using the data in p->data
+    lineHolder->update(vertices, color);
 }
 
 void LaserLine::updateDataIntern(base::samples::LaserScan const& value)
 {
-    p->data = value;
-    std::cout << "got new sample data" << std::endl;
+    loadLineVerticesFromLaserScan(value);
+}
+
+bool LaserLine::isShowRange() const
+{
+    return lineHolder->isShowRange();
+}
+void LaserLine::setShowRange(bool value)
+{
+    lineHolder->setShowRange(value);
+}
+
+QColor LaserLine::getColor()
+{
+    QColor qcolor;
+    qcolor.setRgbF(color[0], color[1], color[2], color[3]);
+    return qcolor;
+}
+
+void LaserLine::setColor(QColor value)
+{
+    if (value.isValid()) {
+        color.set(value.redF(),
+                  value.greenF(),
+                  value.blueF(),
+                  value.alphaF());
+    }
+}
+
+double LaserLine::getLineWidth()
+{
+    return (double)lineHolder->getLineWidth();
+}
+
+void LaserLine::setLineWidth(double lineWidth)
+{
+    lineHolder->setLineWidth((float)lineWidth);
+}
+
+bool LaserLine::getShowAllLines()
+{
+    return showAllLines;
+}
+
+void LaserLine::setShowAllLines(bool value)
+{
+    showAllLines = value;
+}
+
+bool LaserLine::loadLineVerticesFromLaserScan(base::samples::LaserScan const& laserScan)
+{
+    std::vector<uint32_t> ranges = laserScan.ranges;
+
+    double angle = laserScan.start_angle; //start angle
+    double angleStep = laserScan.angular_resolution; //angle step
+    double maxRange = laserScan.maxRange / 1000.0; //convert millimeters to meters
+
+    vertices->clear();
+
+    if (!ranges.empty() && ranges.size() >= 2) {
+
+
+        for (int i = 0; i < ranges.size()-1; i++) {
```

Comment:
Is there some place which data re-computing is happening? Can you explain this solution with more details? What is the problem with the solution right now?